router/connection.go

```diff
@@ -356,6 +380,76 @@ func (conn *LocalConnection) run(actionChan <-chan ConnectionAction, finished ch
 // prevent that completely, since, for example, forwarder can only be
 // created when we know the remote UDP address, but it helps to try.
 
+func (conn *LocalConnection) makeFeatures() map[string]string {
+	return map[string]string{
+		"PeerNameFlavour": PeerNameFlavour,
+		"Name":            conn.local.Name.String(),
+		"NickName":        conn.local.NickName,
+		"UID":             fmt.Sprint(conn.local.UID),
+		"ConnID":          fmt.Sprint(conn.uid),
+	}
+}
+
+type features map[string]string
+
+func (features features) Get(key string) (string, error) {
+	val, ok := features[key]
+	if !ok {
+		return "", fmt.Errorf("Field %s is missing", key)
+	}
+
+	return val, nil
+}
+
+func (conn *LocalConnection) parseFeatures(features features) (*Peer, error) {
+	str, err := features.Get("PeerNameFlavour")
+	if err != nil {
+		return nil, err
+	}
+
+	if str != PeerNameFlavour {
+		return nil, fmt.Errorf("Peer name flavour mismatch (ours: \"%s\", theirs: \"%s\")", PeerNameFlavour, str)
+	}
+
+	str, err = features.Get("Name")
+	if err != nil {
+		return nil, err
+	}
+
+	name, err := PeerNameFromString(str)
+	if err != nil {
+		return nil, err
+	}
+
+	nickName, err := features.Get("NickName")
+	if err != nil {
+		return nil, err
+	}
+
+	str, err = features.Get("UID")
+	if err != nil {
+		return nil, err
+	}
+
+	uid, err := ParsePeerUID(str)
+	if err != nil {
+		return nil, err
+	}
+
+	str, err = features.Get("ConnID")
+	if err != nil {
+		return nil, err
+	}
+
+	remoteConnID, err := strconv.ParseUint(str, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	conn.uid ^= remoteConnID
+	return NewPeer(name, nickName, uid, 0), nil
+}
+
```

Comment:
I have simplified this in https://github.com/rade/weave/commit/ed1efd2539941641af37d775a25dcaa2e948922f